tests: add utility to easily profile node performance with perf

jamesob opened the pull request:
Adds a context manager to easily (and selectively) profile node performance during functional test execution using `perf`. 

While writing some tests, I encountered some odd bitcoind slowness. I wrote up a utility (`TestNode.profile_with_perf`) that generates performance diagnostics for a node by running `perf` during the execution of a particular region of test code.

`perf` usage is detailed in the excellent (and sadly unmerged) https://github.com/bitcoin/bitcoin/pull/12649; all due props to @eklitzke.

### Example

```python
with node.profile_with_perf("large-msgs"):
    for i in range(200):
        node.p2p.send_message(some_large_msg)
    node.p2p.sync_with_ping()
```

This generates a perf data file in the test node's datadir (`/tmp/testtxmpod0y/node0/node-0-TestName-large-msgs.perf.data`).

Running `perf report` generates nice output about where the node spent most of its time while running that part of the test:

```bash
$ perf report -i /tmp/testtxmpod0y/node0/node-0-TestName-large-msgs.perf.data --stdio \
  | c++filt \
  | less

# To display the perf.data header info, please use --header/--header-only options.
#
#
# Total Lost Samples: 0
#
# Samples: 135  of event 'cycles:pp'
# Event count (approx.): 1458205679493582
#
# Children      Self  Command          Shared Object        Symbol
# ........  ........  ...............  ...................  ........................................................................................................................................................................................................................................................................
#
    70.14%     0.00%  bitcoin-net      bitcoind             [.] CNode::ReceiveMsgBytes(char const*, unsigned int, bool&)                                                                                                                                                                                                                        
                |
                ---CNode::ReceiveMsgBytes(char const*, unsigned int, bool&)

    70.14%     0.00%  bitcoin-net      bitcoind             [.] CNetMessage::readData(char const*, unsigned int)                                                                                                                                                                                                                                
                |
                ---CNetMessage::readData(char const*, unsigned int)
                   CNode::ReceiveMsgBytes(char const*, unsigned int, bool&)

    35.52%     0.00%  bitcoin-net      bitcoind             [.] std::vector<char, zero_after_free_allocator<char> >::_M_fill_insert(__gnu_cxx::__normal_iterator<char*, std::vector<char, zero_after_free_allocator<char> > >, unsigned long, char const&)                                                                                      
                |
                ---std::vector<char, zero_after_free_allocator<char> >::_M_fill_insert(__gnu_cxx::__normal_iterator<char*, std::vector<char, zero_after_free_allocator<char> > >, unsigned long, char const&)
                   CNetMessage::readData(char const*, unsigned int)
                   CNode::ReceiveMsgBytes(char const*, unsigned int, bool&)

...
```

jamesob committed:
tests: add utility to easily profile node performance with perf

Introduces `TestNode.profile_with_perf()` context manager which
samples node execution to produce profiling data.

Also introduces a test framework flag, `--perf`, which will run
perf on all nodes for the duration of a given test.

jamesob commented:
An example of running the generated perf data file through [hotspot](https://github.com/KDAB/hotspot):

![selection_109](https://user-images.githubusercontent.com/73197/47235981-b33ebd80-d3a8-11e8-96bd-30f6f91b95f4.png)

ryanofsky commented on test/functional/test_framework/test_node.py:
```diff
@@ -267,6 +267,70 @@ def assert_debug_log(self, expected_msgs):
+                'perf', 'record',
+                '-g',
+                '--call-graph', 'dwarf',
+                '-F', '101',
```
Maybe comment `# Frequency in Hz`

ryanofsky commented on test/functional/test_framework/test_node.py:
```diff
@@ -267,6 +267,70 @@ def assert_debug_log(self, expected_msgs):
+                cmd, shell=True,
+                stderr=subprocess.DEVNULL, stdout=subprocess.DEVNULL).returncode == 0
+
+        if not test_success('readelf -S {} | grep .debug_str'.format(self.binary)):
```
Escaping self.binary with [shlex.quote](https://docs.python.org/3/library/shlex.html#shlex.quote) would let it work with any filename.

ryanofsky reviewed (approved):
utACK b93a17c3153f050eb37292b51463b0aaa5860f1a. Context manager seems nice for writing new tests. This is just an idea, but maybe it would be useful if there were an option to collect perf data for all nodes from start to stop to gain insight into performance of existing tests.

jamesob commented:
Thanks for the review, @ryanofsky. I really like your suggestions and so I've incorporated them all into a rebase.

You can now pass `--perf` when running an individual test file and all nodes will be profiled for the duration of the test. Here's an example run:
```sh
$ ./test/functional/p2p_unrequested_blocks.py --perf

2018-10-19T20:36:06.542000Z TestFramework (INFO): Initializing test directory /tmp/test71_cfvsc
...
2018-10-19T20:36:09.206000Z TestFramework (INFO): Stopping nodes
2018-10-19T20:36:09.384000Z TestFramework.node0 (INFO): Wrote perf output to '/tmp/test71_cfvsc/node0/node-0-AcceptBlockTest.perf.data'
2018-10-19T20:36:09.558000Z TestFramework.node1 (INFO): Wrote perf output to '/tmp/test71_cfvsc/node1/node-1-AcceptBlockTest.perf.data'
2018-10-19T20:36:09.608000Z TestFramework (WARNING): Not cleaning up dir /tmp/test71_cfvsc due to perf data
2018-10-19T20:36:09.608000Z TestFramework (INFO): Tests successful

```

promag commented:
Concept ACK, very nice!

jtimon commented:
Concept ACK

ryanofsky commented on test/functional/test_framework/test_node.py:
```diff
@@ -59,7 +60,15 @@ class TestNode():
+    def __init__(self, i, datadir, *, rpchost, timewait, bitcoind, bitcoin_cli, mocktime, coverage_dir, extra_conf=None, extra_args=None, use_cli=False, start_perf=False, test_class_name=None):
+        """
+        Kwargs:
+            test_class_name (str): The name of the test class which has spawned this node.
```
I don't think I see advantage of the test_class_name plumbing. It seems like it makes resulting filenames less predictable I run a test with --perf and the resulting filenames have a python class name in the middle. It also doesn't seem like it helps with disambiguation either if every node has the same test name.

ryanofsky reviewed (approved):
utACK 358e757d566d5f659901acc6f15fa9eb6eb3e741

jamesob commented:
Addressed feedback from @ryanofsky @conscott.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -267,6 +284,96 @@ def assert_debug_log(self, expected_msgs):
+
+    def _get_perf_data_path(self, profile_name):
+        """Return a suitable path for perf data given some descriptive name."""
+        suffix = "-{}".format(profile_name) if profile_name else ""
```
profile_name is always given, so no need for the else?

jamesob replied:
It's an empty string when we specify `--perf` to the runner: https://github.com/bitcoin/bitcoin/blob/358e757d566d5f659901acc6f15fa9eb6eb3e741/test/functional/test_framework/test_node.py#L184

MarcoFalke replied:
Wouldn't that overwrite the perf file on restart?

jamesob replied:
Ah, good catch. Will verify & fix.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -267,6 +284,96 @@ def assert_debug_log(self, expected_msgs):
+
+        def test_success(cmd):
+            try:
+                return subprocess.run(
```
We don't support this function yet: https://docs.python.org/3.4/library/subprocess.html#subprocess.run

maflcko reviewed (approved):
utACK

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -200,9 +202,18 @@ def main(self):
             self.log.info("Cleaning up {} on exit".format(self.options.tmpdir))
             cleanup_tree_on_exit = True
+        elif self.options.perf:
+            self.log.warning("Not cleaning up dir %s due to perf data" % self.options.tmpdir)
```
nit: Could use `.format(...` like two lines above

jamesob commented:
Pushed feedback from @MarcoFalke.

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -200,9 +202,18 @@ def main(self):
             self.log.info("Cleaning up {} on exit".format(self.options.tmpdir))
             cleanup_tree_on_exit = True
+        elif self.options.perf:
+            self.log.warning("Not cleaning up dir %s due to perf data".format(self.options.tmpdir))
```
needs to use `{}` in the format string ;)

jamesob replied:
Oops! Thanks :)

ryanofsky reviewed (approved):
utACK 5d8ac69be53a876d2b15ba703603cea8f4efdb1a. Just change in string formatting and simplification to file naming since last review.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -167,6 +177,9 @@ def start(self, extra_args=None, *, stdout=None, stderr=None, **kwargs):
         self.log.debug("bitcoind started, waiting for RPC to come up")
 
+        if self.start_perf:
+            self._start_perf("")
```
That will  overwrite previous perf results.

jamesob replied:
Fixed and tested by running `./test/functional/p2p_disconnect_ban.py --perf` (which restarts node1):
```sh
 $ ./test/functional/p2p_disconnect_ban.py --perf
2018-11-09T18:41:23.959000Z TestFramework (INFO): Initializing test directory /tmp/testk5t2fuyc
[...]
2018-11-09T18:41:25.250000Z TestFramework (INFO): setban: test persistence across node restart
2018-11-09T18:41:25.616000Z TestFramework.node1 (INFO): Wrote perf output to '/tmp/testk5t2fuyc/node1/perf.data'
[...]
2018-11-09T18:41:26.438000Z TestFramework (INFO): Stopping nodes
2018-11-09T18:41:26.756000Z TestFramework.node0 (INFO): Wrote perf output to '/tmp/testk5t2fuyc/node0/perf.data'
2018-11-09T18:41:27.040000Z TestFramework.node1 (INFO): Wrote perf output to '/tmp/testk5t2fuyc/node1/1.perf.data'
2018-11-09T18:41:27.041000Z TestFramework (WARNING): Not cleaning up dir /tmp/testk5t2fuyc due to perf data
2018-11-09T18:41:27.041000Z TestFramework (INFO): Tests successful
```

jnewbery commented on test/functional/test_framework/test_node.py:
```diff
@@ -316,6 +333,117 @@ def assert_memory_usage_stable(self, perc_increase_allowed=0.03):
+        Context manager that allows easy profiling of node activity using `perf`.
+
+        Perf will sample the running node and will generate profile data in the node's
+        datadir. The profile data can then be presented using `perf report` or a graphical
```
Since this is now a user activated feature (with `--perf`, I think most of these usage notes should go in `/test/README/md`)

jnewbery reviewed (commented):
Concept ACK.

This seems cool. I haven't had a chance to play around with it yet. I've given the code a cursory glance and will review more thoroughly next week.

jnewbery commented on test/functional/example_test.py:
```diff
@@ -193,10 +193,14 @@ def run_test(self):
-        for block in blocks:
-            getdata_request.inv.append(CInv(2, block))
-        self.nodes[2].p2p.send_message(getdata_request)
+        # If we have Linux perf-tools installed, we can easily profile part of the
```
I think this example code should be in `test/README/md`, which is documentation of how to run tests and contains debugging/troubleshooting tips.

This is meant as an example of how to write a test. I don't think the expectation is that people leave `profile_with_perf` contexts in test that are checked into the repo.

jamesob replied:
The presence of usage here is required to get around vulture's dead code check, but I'll add documentation to `test/functional/README.md`.

jnewbery replied:
If this is just to get around vulture, then I'd prefer to just update `lint-python-dead-code.sh` to ignore the `profile_with_perf` name. This file is supposed to be a guide for writing a new testcase. As far as I can tell, `profile_with_perf` shouldn't be included in new testcases to be checked into master.

jnewbery reviewed (commented):
This is really cool. I've now had a chance to play around with it and it's great. I think doing the following would make it even more useful and accessible for developers:

- move the documentation about installing and using `perf` to `doc/developer-notes.md`
- move the documentation about using `profile_with_perf()` to `test/README.md`, with a reference to `doc/developer-notes.md` for notes on perf.

laanwj commented:
Concept ACK

jamesob commented on test/functional/test_framework/test_node.py:
```diff
@@ -316,6 +333,117 @@ def assert_memory_usage_stable(self, perc_increase_allowed=0.03):
+                cmd, shell=True,
+                stderr=subprocess.DEVNULL, stdout=subprocess.DEVNULL) == 0
+
+        if not test_success('readelf -S {} | grep .debug_str'.format(shlex.quote(self.binary))):
```
Perf isn't available for macOS since it's Linux-specific. I'll add a warning and early-exit for non-Linux systems.

jamesob commented on test/functional/test_framework/test_node.py:
```diff
@@ -18,6 +18,7 @@
 import time
 import urllib.parse
 import collections
+import shlex
```
If we want to enforce alphabetical sorting for imports, we should have a linter for it!

jamesob committed:
docs: add perf section to developer docs

jamesob commented:
Thanks for the review, @jnewbery. I've pushed a rebased tip that addresses the feedback. 

[Here's the diff](https://gist.github.com/jamesob/01ac7b24039621a2863995ef69037a0c) (excluding the rebase).

ryanofsky commented on doc/developer-notes.md:
````diff
@@ -257,6 +258,51 @@ make cov
+```
+
+Make sure you [understand the security
+trade-offs](https://lwn.net/Articles/420403/) of setting these kernel
````
In commit "docs: add perf section to developer docs" (27a895d72a385bcaa861e08b03bb5c817f8e253f)

Slightly strange the link text is "understand the security tradeoffs" instead of just "security tradeoffs."

ryanofsky reviewed (approved):
utACK 27a895d72a385bcaa861e08b03bb5c817f8e253f. Changes since last review: rebase, moving & adding documentation, tweaking output paths.

jnewbery commented on test/functional/README.md:
```diff
@@ -118,3 +118,36 @@ Helpers for script.py
 #### [test_framework/blocktools.py](test_framework/blocktools.py)
 Helper functions for creating blocks and transactions.
+
+### Benchmarking with perf
```
This documentation is excellent. I think it should live in `/test/README.md`, which is documentation for *running* the functional tests, and already includes hints on logging, attaching a debugger, etc. `/test/functional/README.md` is documentation for *writing* test cases (we could probably add a note to the top of each file explaining that).

jnewbery commented on test/functional/test_framework/test_node.py:
```diff
@@ -95,6 +103,8 @@ def __init__(self, i, datadir, *, rpchost, timewait, bitcoind, bitcoin_cli, mock
         self.log = logging.getLogger('TestFramework.node%d' % i)
         self.cleanup_on_exit = True # Whether to kill the node when this object goes away
+        # Cache perf subprocesses here by their data output filename.
+        self.perf_subprocesses = {}
```
My point is that there should only ever be one `perf_subprocess` at any time (it gets popped in `_stop_perf`). I think it's fine to have a `perf_subrocess` reference in the `TestNode` object, but making it a dictionary when there can only be one reference at a time seems odd.

Or maybe I'm misunderstanding. Is it possible to attach more than one perf instance to the same process? Can you give an example of when you'd want that?

jamesob replied:
> My point is that there should only ever be one perf_subprocess at any time

Yeah, this isn't the case: you might have nested context managers, or a context manager used in conjunction with `--perf`.

jnewbery commented on test/functional/test_framework/test_node.py:
```diff
@@ -186,6 +197,9 @@ def start(self, extra_args=None, *, stdout=None, stderr=None, **kwargs):
         self.log.debug("bitcoind started, waiting for RPC to come up")
 
+        if self.start_perf:
+            self._start_perf("")
```
I think it'd be better to pass `None` here (or even better have `None` as the default argument). `""` and `None` will get treated the same in `(profile_name or 'test')`.

EDIT: I see you're using `profile_name` as the key in the `perf_subprocesses` dictionary so this wouldn't work. Is `perf_subprocesses` safe from having entries overwritten if the same `profile_name` is passed in twice?

jnewbery commented on test/functional/test_framework/test_node.py:
```diff
@@ -317,6 +335,84 @@ def assert_memory_usage_stable(self, *, increase_allowed=0.03):
+    def _stop_perf(self, profile_name):
+        """Stop (and pop) a perf subprocess."""
+        subp = self.perf_subprocesses.pop(profile_name)
+        output_path = subp.args[subp.args.index('-o') + 1]
```
This seems a bit brittle. Probably ok to leave as it is, but if anyone updates the arguments or ordering of arguments, then this will break.

jamesob replied:
I don't think this is as brittle as it seems; you're always going to have the filename following `-o` otherwise the argument doesn't make any sense.

jnewbery commented on test/functional/README.md:
```diff
@@ -118,3 +118,36 @@ Helpers for script.py
+
+Perf will sample the running node and will generate profile data in the node's
+datadir. The profile data can then be presented using `perf report` or a graphical
+tool like `hotspot`<https://github.com/KDAB/hotspot>.
```
Make this a markdown link: `[hotspot](https://github.com/KDAB/hotspot).`

jnewbery reviewed (commented):
Looks great James. A few comments inline.

jamesob commented:
Pushed an update addressing feedback from @jnewbery.

ryanofsky reviewed (approved):
utACK 13782b8ba84c5033a59a5234410a763393eafb8d. Main change is removing code in example test, and adding to test README instead.
